crates/storage/src/lib.rs

```diff
@@ -1,163 +1,189 @@
-mod account;
-mod block;
-mod receipt;
-
-use account::{
-    AccountCodeHashRLP, AccountCodeRLP, AccountInfoRLP, AccountStorageKeyRLP,
-    AccountStorageValueRLP, AddressRLP,
-};
-use block::{BlockBodyRLP, BlockHeaderRLP};
-use ethereum_rust_core::types::{BlockNumber, Index};
-use libmdbx::{
-    dupsort,
-    orm::{table, Database},
-    table_info,
-};
-use receipt::ReceiptRLP;
-use std::path::Path;
-
-// Define tables
-table!(
-    /// Block headers table.
-    ( Headers ) BlockNumber => BlockHeaderRLP
-);
-table!(
-    /// Block bodies table.
-    ( Bodies ) BlockNumber => BlockBodyRLP
-);
-table!(
-    /// Account infos table.
-    ( AccountInfos ) AddressRLP => AccountInfoRLP
-);
-dupsort!(
-    /// Account storages table.
-    ( AccountStorages ) AddressRLP[AccountStorageKeyRLP] => AccountStorageValueRLP
-);
-table!(
-    /// Account codes table.
-    ( AccountCodes ) AccountCodeHashRLP => AccountCodeRLP
-);
-dupsort!(
-    /// Receipts table.
-    ( Receipts ) BlockNumber[Index] => ReceiptRLP
-);
-
-/// Initializes a new database with the provided path. If the path is `None`, the database
-/// will be temporary.
-pub fn init_db(path: Option<impl AsRef<Path>>) -> Database {
-    let tables = [
-        table_info!(Headers),
-        table_info!(Bodies),
-        table_info!(AccountInfos),
-        table_info!(AccountStorages),
-        table_info!(AccountCodes),
-        table_info!(Receipts),
-    ]
-    .into_iter()
-    .collect();
-    let path = path.map(|p| p.as_ref().to_path_buf());
-    Database::create(path, &tables).unwrap()
+//! Abstraction for persistent data storage.
+//!    Supporting InMemory and Libmdbx storage.
+//!    There is also a template for Sled and RocksDb implementation in case we
+//!    want to test or benchmark against those engines (Currently disabled behind feature flags
+//!    to avoid requiring the implementation of the full API).
+
+#[cfg(feature = "in_memory")]
+use self::in_memory::Store as InMemoryStore;
+#[cfg(feature = "libmdbx")]
+use self::libmdbx::Store as LibmdbxStore;
+#[cfg(feature = "rocksdb")]
+use self::rocksdb::Store as RocksDbStore;
+#[cfg(feature = "sled")]
+use self::sled::Store as SledStore;
+use anyhow::Result;
+use ethereum_rust_core::types::AccountInfo;
+use ethereum_types::Address;
+use std::fmt::Debug;
+use std::sync::{Arc, Mutex};
+
+mod rlp;
+
+#[cfg(feature = "in_memory")]
+mod in_memory;
+#[cfg(feature = "libmdbx")]
+mod libmdbx;
+#[cfg(feature = "rocksdb")]
+mod rocksdb;
+#[cfg(feature = "sled")]
+mod sled;
+
+pub(crate) type Key = Vec<u8>;
+pub(crate) type Value = Vec<u8>;
+
+pub trait StoreEngine: Debug + Send {
+    fn add_account_info(&mut self, address: Address, account_info: AccountInfo) -> Result<()>;
+    fn get_account_info(&self, address: Address) -> Result<Option<AccountInfo>>;
+    fn set_value(&mut self, key: Key, value: Value) -> Result<()>;
+    fn get_value(&self, key: Key) -> Result<Option<Value>>;
+}
```

Comment:
We should add some documentation on the trait, or at least the `set_value`/`get_value` methods.